### PR TITLE
fix(backfill): stop re-fetching blocks evicted by the recent-tier retention policy

### LINKED ISSUE #2698: Bug: Backfilled Blocks Immediately Evicted by Recent Storage Retention Policy
# Bug: Backfilled Blocks Immediately Evicted by Recent Storage Retention Policy

## Summary

When running the **remote plugins profile** (recent-file storage + S3 archive, no historical storage), backfill detects historical gaps and fetches old blocks from peers. Those blocks land in recent storage, which immediately deletes them because they are the oldest entries in the retention window. S3 never archives them. The next gap scan re-detects the same gap, creating an infinite retry loop.

## Impact

Historical gaps that fall outside the recent retention window can **never be permanently stored** in this configuration. S3 is left with a permanent gap matching the missed blocks. Backfill wastes peer bandwidth retrying indefinitely.

### Steps to Reproduce

1. Start a block node with the remote plugins profile and default retention (96,000 blocks).
2. Let recent storage fill to capacity (receive ≥ 96,001 live blocks).
3. Create a historical gap: stop the node before the window was full, miss a range of blocks, then restart so the node continues from a later point, leaving a gap below `latest - 96,000`. _Alternatively_: configure `startBlock` in backfill to point at a gap that is older than the current retention window.
4. Enable backfill pointing at a peer that has the missing blocks.
5. **Observe:**
   - Backfill logs `Backfilling block [N]` for the gap block.
   - Retention immediately deletes block N (check recent-storage metrics or log at DEBUG).
   - S3 never receives an upload for block N.
   - The gap re-appears in the next backfill scan cycle (infinite loop).

## Root Cause

Recent storage (`BlockFileRecentPlugin.handleVerification`) runs the retention policy in the **same event handler call** that writes a newly received block:

```
writeBlockToLivePath(block)          // block is now the OLDEST entry
→ excess = availableBlocks.size() - 96_000
→ delete from availableBlocks.min()  // min() == the block we just wrote
```

Because a historical gap is, by definition, older than all currently stored blocks, the backfilled block becomes `availableBlocks.min()` the instant it is written. With a full 96,000-block window, `excess == 1` and the block is deleted immediately.

S3 archive checks `available.contains(startBlock, endBlock)` before queuing an upload (`S3ArchivePlugin.java:224`). The window between "block written" and "block deleted" is a single event-handler pass — effectively zero from S3's perspective.

## Affected Configuration

- **Profile:** remote plugins (recent-file storage + S3 archive)
- **Setting:** `blockRetentionThreshold > 0` (default 96,000)
- **Condition:** any gap whose block number is below `latestBlock - 96,000`

### LINKED ISSUE #2820: Update Backfill to use ApplicationStateFacility
### Story Form

As a BlockNode engineer I would like the ApplicationStateFacility to be the source of truth for all block range searches.

### Technical Notes

Backfill currently searches gaps from block providers and should switch to pulling from new logic in applicationStateFacility

### PR DESCRIPTION
**Pull request by AlfredoG87**
fix(backfill): stop re-fetching blocks evicted by the recent-tier retention policy

## Problem

On a recent-tier deployment with retention, backfill fetched a historical block, retention evicted it, and the next scan re-detected the now-missing block and re-fetched it — forever, wasting peer bandwidth and local churn (#2698). The root cause: backfill detected gaps from *available* blocks (what's retrievable right now) and consumed only a peer's fuzzy `first/last` range.

## Change

Backfill is now a pure consumer of Application State: it reads block-range state and never writes it. The stored set is owned by the storage/archive plugins (cloud-storage-archive) and reflects what was actually persisted remotely.

- **Gap detection uses stored ∪ available.** A block a storage plugin recorded as stored is not re-backfilled after the recent tier evicts it from `available`. Retention only removes from the recent plugin's own `available` set; the stored set is monotonic and controlled elsewhere. The plugin caches the stored+available snapshot delivered via `onContextUpdate()` (the same merge `BlockNodeApp` already computes when the underlying data changes) and unions it with a live `availableBlocks()` read on every scan, so it never re-queries the Application State facility's raw stored set per scan while still reflecting blocks persisted or evicted since the last context update.

- **Exact peer ranges via `serverStatusDetail`.** Replaces the coarse `serverStatus` first/last, so peers with internal gaps are handled precisely and a chunk is only requested from a peer that actually holds it. `getNewAvailableRanges` now returns the union of peers' real ranges instead of fabricating a single contiguous span.

- **Exponential backoff for reappearing gaps.** When a gap keeps reappearing without shrinking (e.g. recent-only deployments where nothing records the block as stored), it is retried with an exponentially growing delay (base = scan interval, capped at `maxBackoffMs`) instead of every scan. Backoff resets the moment a gap makes real progress.

Net effect: with a cloud archive the loop simply cannot happen (archived blocks stay in `stored`, so they're never re-detected). Without one, backfill still won't hammer — it backs off to the cap.

## Notable

- `BackfillPlugin` now implements `onContextUpdate()` to cache the merged stored+available block ranges reported by the Application State facility, instead of adding a new read accessor to `ApplicationStateFacility` — the initial approach was flagged in review as inconsistent with the already-merged `BlockNodeContext.storedBlocks()` and as redundant per-scan work.
- Fixed a bug in the `TestBlockNodeServer` fixture where `serverStatusDetail` reported only the last available range.
- The old "lying node marked unavailable" test now reflects the new behavior: with exact ranges a peer that has blocks on both sides of a gap serves both.
- `PluginTestBase` gained `updateStoredBlocks()`, which records a stored range and delivers it to the plugin under test via `onContextUpdate()`, matching how `BlockNodeApp` merges stored and available blocks in production.
- Fixed a pre-existing flaky test (`BlockNodeAppTest.testBlockRangesTriggerOnContextUpdate`) that assumed a fixed number of `onContextUpdate` calls; two rapid state changes can coalesce into a single scanner tick, so the test now waits on the delivered content rather than a call count.

## Testing

Tests in `BackfillPluginTest` cover: stored blocks are not re-backfilled, a reappearing gap is throttled by backoff (not re-fetched every scan), and exact peer ranges are consumed while the peer's internal gap is skipped. `BackfillFetcherTest` mocks moved to `serverStatusDetail`. All affected module suites (`spi-plugins`, `app`, `backfill`, `cloud-storage-archive`, `health`, `server-status`, `block-verification`) and `qualityCheck` pass.

## Related issues

Closes #2698, closes #2820, advances #2819.

### CODE REVIEW

**Review by berryware — DISMISSED** [review]
Looks good.

File: block-node/app/src/main/java/org/hiero/block/node/app/BlockNodeApp.java [hunk 1]
```diff
@@ -499,6 +499,11 @@ public void addStoredBlockRange(LongRange blockRange) {
         storedBlocks.add(blockRange);
     }
 
+    @Override
+    public BlockRangeSet storedBlocks() {
+        return storedBlocks;
+    }
+
```
Comment (jsync-swirlds) [inline block-node/app/src/main/java/org/hiero/block/node/app/BlockNodeApp.java]: The problem with a trivial accessor here is that it will not be consistent with the latest Context; because the stored blocks are merged with available blocks on context update, while this is just the blocks explicitly labeled as `stored`.

File: block-node/app/src/test/java/org/hiero/block/node/app/BlockNodeAppTest.java [hunk 2]
```diff
@@ -1044,16 +1044,25 @@ void testBlockRangesTriggerOnContextUpdate() throws Exception {
...
+            final BlockNodeContext context = testPlugin.getContext();
+            if (context != null && !context.storedBlocks().isEmpty()) {
+                storedBlocks = context.storedBlocks().getFirst();
+                if (storedBlocks.rangeStart() == 0L && storedBlocks.rangeEnd() == 1049L) {
+                    break;
+                }
+            }
+            //noinspection BusyWait
+            Thread.sleep(50);
+        }
```
Comment (jsync-swirlds) [inline block-node/app/src/test/java/org/hiero/block/node/app/BlockNodeAppTest.java]: The await used a condition wait, which is more reliable and _even that_ still fails sometimes.
We need to be improving test reliability, not adding more spin waits that sometimes work and sometimes fail.

Is there a mechanism we could be using that either uses a controllable execution service so there is no need to guess, or has an unambiguous condition wait?

File: block-node/backfill/src/main/java/org/hiero/block/node/backfill/BackfillFetcher.java [hunk 3]
```diff
@@ -130,14 +131,12 @@ public LongRange getNewAvailableRange(long latestStoredBlockNumber) {
...
-
-                // update the earliestPeerBlock to the max lastAvailableBlock
-                latestPeerBlock = Math.max(latestPeerBlock, lastAvailableBlock);
-                earliestPeerBlock = Math.min(earliestPeerBlock, firstAvailableBlock);
+                for (final LongRange range : resolveAvailableRanges(currentNodeClient)) {
+                    if (range.end() > latestStoredBlockNumber) {
+                        final long start = Math.max(range.start(), latestStoredBlockNumber + 1);
+                        peerRanges.add(new LongRange(start, range.end()));
+                    }
+                }
```
Comment (jsync-swirlds) [inline block-node/backfill/src/main/java/org/hiero/block/node/backfill/BackfillFetcher.java:144]: This seems off.
Why are we modifying the start of ranges based on the highest stored block; are we deciding that backfill will never fill in gaps below that point?

File: block-node/backfill/src/main/java/org/hiero/block/node/backfill/BackfillPlugin.java [hunk 4]
```diff
@@ -296,11 +314,18 @@ private void shutdownExecutor(ExecutorService executor, String name) {
...
+        //    evicted from a volatile tier (e.g. by the recent-tier retention policy) but were already stored.
+        ConcurrentLongRangeSet knownBlocks = new ConcurrentLongRangeSet();
+        knownBlocks.addAll(
+                context.applicationStateFacility().storedBlocks().streamRanges().toList());
+        knownBlocks.addAll(context.historicalBlockProvider()
                 .availableBlocks()
                 .streamRanges()
-                .toList();
+                .toList());
+        List<LongRange> blockRanges = knownBlocks.streamRanges().toList();
```
Comment (jsync-swirlds) [inline block-node/backfill/src/main/java/org/hiero/block/node/backfill/BackfillPlugin.java]: The ASF does this work when updating the context.
It just does so only when the data is meaningfully updated, rather than every time the data is queried.

Not sure which approach has better performance, but we know that performance is already a mild issue.

File: block-node/spi-plugins/src/main/java/org/hiero/block/node/spi/ApplicationStateFacility.java [hunk 5]
```diff
@@ -47,6 +48,20 @@ default boolean updateAddressBookHistory(RangedAddressBookHistory history) {
...
+    /// are not re-fetched.
+    ///
+    /// The default implementation returns an empty set. Implementations that track stored blocks
+    /// (i.e. `BlockNodeApp`) override this method.
+    ///
+    /// @return the set of stored blocks; never `null` (empty when none are stored)
+    default BlockRangeSet storedBlocks() {
+        return BlockRangeSet.EMPTY;
+    }
+
```
Comment (jsync-swirlds) [inline block-node/spi-plugins/src/main/java/org/hiero/block/node/spi/ApplicationStateFacility.java]: I'm not a fan of default implementations.
It hides places we forgot to update and really should have.

**Review by jsync-swirlds — COMMENTED** [review]
A few concerns worth noting; all are advisory.

File: block-node/backfill/src/main/java/org/hiero/block/node/backfill/BackfillPlugin.java (hunk 4 above)
Comment (ata-nas) [reply block-node/backfill/src/main/java/org/hiero/block/node/backfill/BackfillPlugin.java]: I think we should be listening for updates instead of calling a method `storedBlock`, like in other places, and keep a local copy of the latest update. Then, we keep a local copy of those which can be served.

File: block-node/app/src/main/java/org/hiero/block/node/app/BlockNodeApp.java (hunk 1 above)
Comment (AlfredoG87) [reply block-node/app/src/main/java/org/hiero/block/node/app/BlockNodeApp.java]: Removed the trivial storedBlocks() accessor since it wasn't consistent with the already-merged BlockNodeContext.storedBlocks() ✅

File: block-node/spi-plugins/src/main/java/org/hiero/block/node/spi/ApplicationStateFacility.java (hunk 5 above)
Comment (AlfredoG87) [reply block-node/spi-plugins/src/main/java/org/hiero/block/node/spi/ApplicationStateFacility.java]: Removed the new default storedBlocks() method entirely instead of keeping a default implementation

✅

File: block-node/backfill/src/main/java/org/hiero/block/node/backfill/BackfillPlugin.java (hunk 4 above)
Comment (AlfredoG87) [reply block-node/backfill/src/main/java/org/hiero/block/node/backfill/BackfillPlugin.java]: Replaced the per-scan ASF query + re-merge with a cached stored+available snapshot updated via onContextUpdate(), union with a live availableBlocks() read each scan ✅

File: block-node/backfill/src/main/java/org/hiero/block/node/backfill/BackfillFetcher.java (hunk 3 above)
Comment (AlfredoG87) [reply block-node/backfill/src/main/java/org/hiero/block/node/backfill/BackfillFetcher.java:144]: Clarified via javadoc that the clamped range only sizes the greedy scan window and doesn't gate which gaps below baseline get backfilled (that's handled separately by getAvailabilityForRange)

File: block-node/app/src/test/java/org/hiero/block/node/app/BlockNodeAppTest.java (hunk 2 above)
Comment (AlfredoG87) [reply block-node/app/src/test/java/org/hiero/block/node/app/BlockNodeAppTest.java]: Replaced the busy-wait polling loop with a BlockingQueue-drained condition wait (awaitContext) that doesn't assume a fixed update count ✅

File: block-node/backfill/src/main/java/org/hiero/block/node/backfill/BackfillPlugin.java [hunk 6]
```diff
@@ -95,6 +95,13 @@ public class BackfillPlugin implements BlockNodePlugin, BlockNotificationHandler
     // Deduplication: highest block scheduled for live-tail (prevents overlapping submissions)
     private final AtomicLong liveTailHighWaterMark = new AtomicLong(-1);
 
+    // Cached stored+available snapshot delivered by the last onContextUpdate() call. The Application
+    // State facility computes this merge itself whenever the underlying data actually changes, so
+    // caching it here avoids re-querying its raw stored set on every autonomous scan. Each scan still
+    // unions this with a live availableBlocks() read (see detectAndScheduleGaps) to reflect blocks
+    // persisted or evicted since the last context update.
+    private volatile List<LongRange> knownBlockRanges = List.of();
```
Comment (jsync-swirlds) [inline block-node/backfill/src/main/java/org/hiero/block/node/backfill/BackfillPlugin.java]: Instead of a List, perhaps a `ConcurrentLongRangeSet` would be better?
That would make merging the ranges (to combine overlapping and contiguous ranges) easier.

This works even better if we add a `List` accessor to `ConcurrentLongRangeSet` (possibly returning an unmodifiable view using `Collections.unmodifiableList`) so we don't need to constantly remap via a stream just to get a List.

File: block-node/backfill/src/main/java/org/hiero/block/node/backfill/BackfillPlugin.java [hunk 7]
```diff
@@ -187,6 +194,16 @@ public void init(BlockNodeContext context, ServiceBuilder serviceBuilder) {
...
     }
 
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public void onContextUpdate(BlockNodeContext context) {
+        knownBlockRanges = context.storedBlocks().stream()
+                .map(range -> new LongRange(range.rangeStart(), range.rangeEnd()))
+                .toList();
```
Comment (jsync-swirlds) [inline block-node/backfill/src/main/java/org/hiero/block/node/backfill/BackfillPlugin.java]: Why are we remapping what is already a List?
`knownBlockRanges = context.storedBlocks()` would be equivalent, and avoids a streamed remap that doesn't need to happen...

File: block-node/backfill/src/main/java/org/hiero/block/node/backfill/BackfillPlugin.java
Comment (jsync-swirlds) [inline block-node/backfill/src/main/java/org/hiero/block/node/backfill/BackfillPlugin.java]: Grrr...
We should have an `addAll` override on `ConcurrentLongRangeSet` that takes another `ConcurrentLongRangeSet` so we're not having to perform this conversion all the time.

**Review by jsync-swirlds — DISMISSED** [review]
A few more potential improvements.
These can be done in a follow-up if desired.

File: block-node/backfill/src/main/java/org/hiero/block/node/backfill/BackfillPlugin.java (hunk 7 above)
Comment (AlfredoG87) [reply block-node/backfill/src/main/java/org/hiero/block/node/backfill/BackfillPlugin.java]: knownBlockRanges is now a ConcurrentLongRangeSet built directly from context.storedBlocks(), no more mapping into List<LongRange>  ✅

File: block-node/backfill/src/main/java/org/hiero/block/node/backfill/BackfillPlugin.java
Comment (AlfredoG87) [reply block-node/backfill/src/main/java/org/hiero/block/node/backfill/BackfillPlugin.java]: Widened addAll(ConcurrentLongRangeSet) to addAll(BlockRangeSet) (a strict superset, so all existing callers still compile) — the merge sites in detectAndScheduleGaps/handleNewestBlockKnownToNetwork now call addAll/max() straight on the cache and on availableBlocks(), no conversions left ✅

File: block-node/backfill/src/main/java/org/hiero/block/node/backfill/BackfillPlugin.java (hunk 6 above)
Comment (AlfredoG87) [reply block-node/backfill/src/main/java/org/hiero/block/node/backfill/BackfillPlugin.java]: Made knownBlockRanges a ConcurrentLongRangeSet and added a toList() accessor on ConcurrentLongRangeSet that returns the internal immutable list directly (no stream remap needed) ✅

**Review by ivannov — DISMISSED** [review]

File: block-node/backfill/src/main/java/org/hiero/block/node/backfill/BackfillPlugin.java [hunk 8]
```diff
@@ -296,11 +332,16 @@ private void shutdownExecutor(ExecutorService executor, String name) {
...
-                .toList();
+        // 1. Get the blocks this node already has: the cached stored+available snapshot from the last
+        //    onContextUpdate() (avoids re-querying the Application State facility's raw stored set on
+        //    every scan) unioned with the live available set, so blocks persisted or evicted since the
+        //    last context update are reflected immediately. Folding in the cached snapshot — not just
+        //    live available — prevents re-backfilling blocks that have been evicted from a volatile tier
+        //    (e.g. by the recent-tier retention policy) but were already stored.
+        ConcurrentLongRangeSet knownBlocks = new ConcurrentLongRangeSet();
+        knownBlocks.addAll(knownBlockRanges);
+        knownBlocks.addAll(context.historicalBlockProvider().availableBlocks());
```
Comment (ata-nas) [inline block-node/backfill/src/main/java/org/hiero/block/node/backfill/BackfillPlugin.java:349]: Why are we adding the available blocks from historical provider? Stored blocks should already be a superset of that. This seems off.

File: block-node/app/src/testFixtures/java/org/hiero/block/node/app/fixtures/plugintest/PluginTestBase.java [hunk 9]
```diff
@@ -363,10 +363,9 @@ public void addStoredBlockRange(LongRange blockRange) {
     public void updateStoredBlocks(LongRange blockRange) {
         appStoredBlocks.add(blockRange);
         final ConcurrentLongRangeSet merged = new ConcurrentLongRangeSet();
-        merged.addAll(appStoredBlocks.streamRanges().toList());
-        merged.addAll(
-                activeHistoricalBlockFacility.availableBlocks().streamRanges().toList());
-        final List<BlockRange> mergedRanges = merged.streamRanges()
+        merged.addAll(appStoredBlocks);
+        merged.addAll(activeHistoricalBlockFacility.availableBlocks());
+        final List<BlockRange> mergedRanges = merged.toList().stream()
```
Comment (jsync-swirlds) [inline block-node/app/src/testFixtures/java/org/hiero/block/node/app/fixtures/plugintest/PluginTestBase.java]: This is running two unnecessary conversions.
Use `streamRanges` instead of converting to a list, then streaming, then remapping back to a range.

File: block-node/backfill/src/main/java/org/hiero/block/node/backfill/BackfillPlugin.java [hunk 10]
```diff
@@ -169,6 +195,16 @@ public void init(BlockNodeContext context, ServiceBuilder serviceBuilder) {
         context.blockMessaging().registerBlockNotificationHandler(this, false, "BackfillPlugin");
     }
 
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public void onContextUpdate(BlockNodeContext context) {
+        final ConcurrentLongRangeSet updated = new ConcurrentLongRangeSet();
+        context.storedBlocks().forEach(range -> updated.add(range.rangeStart(), range.rangeEnd()));
```
Comment (jsync-swirlds) [inline block-node/backfill/src/main/java/org/hiero/block/node/backfill/BackfillPlugin.java]: This is inefficient (stream actions are always more costly than the equivalents) and somewhat confusing (it looks like you're updating context, rather than adding context into "updated".

Why not use addAll?

```suggestion
        updated.addAll(context.storedBlocks());
```

File: block-node/backfill/src/main/java/org/hiero/block/node/backfill/BackfillPlugin.java (hunk 8 above)
Comment (jsync-swirlds) [reply block-node/backfill/src/main/java/org/hiero/block/node/backfill/BackfillPlugin.java:349]: This is merging in updates to historical that happened after the latest update to stored (which is much less frequent, in most cases).

File: block-node/base/src/main/java/org/hiero/block/node/base/ranges/ConcurrentLongRangeSet.java [hunk 11]
```diff
@@ -219,13 +219,10 @@ public final void addAll(Collection<LongRange>... ranges) {
...
-     * @param ranges the ranges to add
+     * @param ranges the set of ranges to add
      */
-    public void addAll(ConcurrentLongRangeSet ranges) {
-        List<LongRange> otherRanges = ranges.ranges.get();
-        for (LongRange range : otherRanges) {
-            add(range);
-        }
+    public void addAll(BlockRangeSet ranges) {
+        ranges.streamRanges().forEach(this::add);
```
Comment (jsync-swirlds) [inline block-node/base/src/main/java/org/hiero/block/node/base/ranges/ConcurrentLongRangeSet.java]: Please avoid streams and lambdas in hot loops like this method can be.
The for loop avoids at least two vcall overheads and two object allocations compared to this construction.

File: block-node/base/src/main/java/org/hiero/block/node/base/ranges/ConcurrentLongRangeSet.java [hunk 12]
```diff
@@ -554,6 +551,15 @@ public Stream<LongRange> streamRanges() {
         return ranges.get().stream();
     }
 
+    /**
+     * Returns the ranges in the set as an unmodifiable list, without the overhead of streaming.
+     *
+     * @return an unmodifiable, ascending, non-overlapping list of the ranges in the set
+     */
+    public List<LongRange> toList() {
+        return ranges.get();
```
Comment (jsync-swirlds) [inline block-node/base/src/main/java/org/hiero/block/node/base/ranges/ConcurrentLongRangeSet.java]: We need to protect the list we return to avoid unintended modifications.


```suggestion
        return Collections.unmodifiableList(ranges.get());
```

File: block-node/backfill/src/main/java/org/hiero/block/node/backfill/BackfillPlugin.java (hunk 8 above)
Comment (AlfredoG87) [reply block-node/backfill/src/main/java/org/hiero/block/node/backfill/BackfillPlugin.java:349]: You're right — removed the live availableBlocks() merge entirely. not really needed, great catch! ✅

File: block-node/app/src/testFixtures/java/org/hiero/block/node/app/fixtures/plugintest/PluginTestBase.java (hunk 9 above)
Comment (AlfredoG87) [reply block-node/app/src/testFixtures/java/org/hiero/block/node/app/fixtures/plugintest/PluginTestBase.java]: Fixed ✅

File: block-node/backfill/src/main/java/org/hiero/block/node/backfill/BackfillPlugin.java (hunk 10 above)
Comment (AlfredoG87) [reply block-node/backfill/src/main/java/org/hiero/block/node/backfill/BackfillPlugin.java]: Replaced the forEach lambda with a plain for loop ✅

File: block-node/base/src/main/java/org/hiero/block/node/base/ranges/ConcurrentLongRangeSet.java (hunk 11 above)
Comment (AlfredoG87) [reply block-node/base/src/main/java/org/hiero/block/node/base/ranges/ConcurrentLongRangeSet.java]: Restored addAll(ConcurrentLongRangeSet) as a plain for loop over the internal list (zero-allocation fast path), and kept addAll(BlockRangeSet) as a separate overload for the generic case, also implemented as a for loop instead of streamRanges().forEach(this::add). Callers passing a ConcurrentLongRangeSet-typed reference get the fast overload automatically 

✅

File: block-node/base/src/main/java/org/hiero/block/node/base/ranges/ConcurrentLongRangeSet.java (hunk 12 above)
Comment (AlfredoG87) [reply block-node/base/src/main/java/org/hiero/block/node/base/ranges/ConcurrentLongRangeSet.java]: Applied as suggested ✅

Comment (jsync-swirlds) [reply block-node/base/src/main/java/org/hiero/block/node/base/ranges/ConcurrentLongRangeSet.java]: Might be this did not get pushed?

File: block-node/backfill/src/main/java/org/hiero/block/node/backfill/BackfillPlugin.java (hunk 8 above)
Comment (jsync-swirlds) [reply block-node/backfill/src/main/java/org/hiero/block/node/backfill/BackfillPlugin.java:349]: Seems there was an issue when removing this?

Comment (AlfredoG87) [reply block-node/backfill/src/main/java/org/hiero/block/node/backfill/BackfillPlugin.java:349]: correct, I had to roll it back since it was failing e2e tests like the cached value is not reliable in practice, needs more investigation and out of scope for this PR, but will create a follow-up issue.

Comment (ata-nas) [reply block-node/backfill/src/main/java/org/hiero/block/node/backfill/BackfillPlugin.java:349]: Yeah, I feel like when you get stored blocks from ASF, we have already merged all available at the point of the call, inside ASF. I agree with @AlfredoG87 We have to investigate as to why we see failures.

File: block-node/backfill/src/main/java/org/hiero/block/node/backfill/BackfillPlugin.java [hunk 13]
```diff
@@ -321,13 +366,53 @@ private void detectAndScheduleGaps() {
...
+                final long start = gap.range().start();
+                detectedHistoricalStarts.add(start);
+                if (isHistoricalGapInBackoff(start, gap.range().size(), now)) {
+                    final String backingOffMsg = "Historical gap [{0}] is backing off, skipping this scan";
+                    LOGGER.log(DEBUG, backingOffMsg, gap.range());
+                    continue;
+                }
+            }
+            if (scheduleGap(gap)) {
+                metricsHolder.backfillGapsDetected.increment();
```
Comment (Nana-EC) [inline block-node/backfill/src/main/java/org/hiero/block/node/backfill/BackfillPlugin.java]: Q: How will this impact metrics inferences?
Before this was incremented on every detection. Now, it's only incremented if that detection is also submitted for backfill.
Either the metric name and description need to be refined or we restore it so the intent is unchanged.
Let's validate to ensure the inferences from metrics don't change, especially if we're alerting on it.

File: block-node/app/src/test/java/org/hiero/block/node/app/BlockNodeAppTest.java [hunk 14]
```diff
@@ -336,6 +340,28 @@ void awaitContextUpdates(final long timeoutSeconds) throws InterruptedException
...
+         * next tick.
+         *
+         * @return the first satisfying context, or {@code null} if the timeout elapses first
+         */
+        BlockNodeContext awaitContext(final long timeoutSeconds, final Predicate<BlockNodeContext> condition)
+                throws InterruptedException {
+            final long deadlineNanos = System.nanoTime() + TimeUnit.SECONDS.toNanos(timeoutSeconds);
+            long remainingNanos;
+            BlockNodeContext delivered;
+            while ((remainingNanos = deadlineNanos - System.nanoTime()) > 0
```
Comment (Nana-EC) [inline block-node/app/src/test/java/org/hiero/block/node/app/BlockNodeAppTest.java]: Should this be >= or should we drop the outer poll?
Considering the edge case of a really small remainingNanos e.g. ~1ns
I'm not sure on the impact of testing but watching out for flakiness here

**Review by Nana-EC — DISMISSED** [review]
No blocking comments.
Let's get this in to unblock RFH and take updates in sprint starting today

File: block-node/backfill/src/main/java/org/hiero/block/node/backfill/BackfillPlugin.java (hunk 13 above)
Comment (AlfredoG87) [reply block-node/backfill/src/main/java/org/hiero/block/node/backfill/BackfillPlugin.java]: Good Catch!! Restored backfill_gaps_detected to its original unconditional semantics (matches existing docs/dashboard, no alerting change) and added a new backfill_gaps_submitted counter for the backoff-throttled case, which the backoff test now asserts on instead. Updated both metrics docs. ✅

File: block-node/app/src/test/java/org/hiero/block/node/app/BlockNodeAppTest.java (hunk 14 above)
Comment (AlfredoG87) [reply block-node/app/src/test/java/org/hiero/block/node/app/BlockNodeAppTest.java]: Yes you are correct! Dropped the outer check, poll() already treats timeout ≤ 0 as non-blocking and returns immediately, so we don't need to reason about > vs >= at all. ✅

File: block-node/base/src/main/java/org/hiero/block/node/base/ranges/ConcurrentLongRangeSet.java (hunk 12 above)
Comment (AlfredoG87) [reply block-node/base/src/main/java/org/hiero/block/node/base/ranges/ConcurrentLongRangeSet.java]: Yes, but it should be there now 👍

File: block-node/app/src/main/java/org/hiero/block/node/app/BlockNodeApp.java [hunk 15]
```diff
@@ -667,7 +667,7 @@ private boolean updateBlockNodeContext(
             addressBookHistory = null;
         }
 
-        List<BlockRange> storedBlockRange = toBlockRange(storedBlocks);
+        List<BlockRange> storedBlockRange = mergeRanges(storedBlocks, availableBlocks);
```
Comment (jsync-swirlds) [inline block-node/app/src/main/java/org/hiero/block/node/app/BlockNodeApp.java:670]: We shouldn't merge ranges unless the available ranges changed.
Otherwise we spend a very expensive merge when we do not need to, and also have a higher chance of updating stored when we should not need to.

File: block-node/backfill/src/main/java/org/hiero/block/node/backfill/BackfillPlugin.java [hunk 16]
```diff
@@ -91,6 +98,27 @@ public class BackfillPlugin implements BlockNodePlugin, BlockNotificationHandler
     // Deduplication: highest block scheduled for live-tail (prevents overlapping submissions)
     private final AtomicLong liveTailHighWaterMark = new AtomicLong(-1);
 
+    // Cached stored+available snapshot delivered by the last onContextUpdate() call. The Application
+    // State facility already folds availableBlocks() into storedBlocks whenever either changes, so this
+    // alone is the source of truth for "blocks this node already has" — no separate live availableBlocks()
+    // read is needed here. A fresh set replaces this reference on each update rather than being mutated
+    // in place, so concurrent readers always see a consistent snapshot.
+    private volatile ConcurrentLongRangeSet knownBlockRanges = new ConcurrentLongRangeSet();
```
Comment (jsync-swirlds) [inline block-node/backfill/src/main/java/org/hiero/block/node/backfill/BackfillPlugin.java:106]: Are we expecting to set this reference to a new value often?
Volatile has no effect on the content of the object here, it only applies to the reference itself.

Basically it doesn't protect the actual content in any way (which is fine if that's intended).

File: block-node/backfill/src/main/java/org/hiero/block/node/backfill/BackfillPlugin.java [hunk 17]
```diff
@@ -296,11 +336,18 @@ private void shutdownExecutor(ExecutorService executor, String name) {
...
+        //    onContextUpdate() unioned with a live availableBlocks() read. The Application State
+        //    facility's own periodic merge normally makes the cache alone sufficient, but its scan is
+        //    on a separate schedule and can (rarely) observe a transient/incomplete snapshot of the
+        //    underlying facility right as it's queried; re-checking the live set here is a cheap
+        //    guard against acting on such a stale snapshot. Folding in stored — not just available —
+        //    prevents re-backfilling blocks that have been evicted from a volatile tier (e.g. by the
+        //    recent-tier retention policy) but were already stored.
+        ConcurrentLongRangeSet knownBlocks = new ConcurrentLongRangeSet();
+        knownBlocks.addAll(knownBlockRanges);
+        knownBlocks.addAll(context.historicalBlockProvider().availableBlocks());
```
Comment (jsync-swirlds) [inline block-node/backfill/src/main/java/org/hiero/block/node/backfill/BackfillPlugin.java:349]: I suspect this is needed because historical blocks facility updates itself much more often than context is updated.

**Review by jsync-swirlds — APPROVED** [review]
Looks good, one question we can address later, and a couple items to note.

You're welcome to resolve all three once read.

**Review by Nana-EC — APPROVED** [review]
LG
